### DISCUSSION TITLE
dependable objects - best practice?

### DISCUSSION THREAD

**Discussion by GergelyGabor** [Snowflake-Labs/schemachange#371] [Ideas]
Hi all, 
I'm implementing Schemachange for our use case, and wanted to deploy views as Repeatable scripts. 
Given that this is a new project, and we are probably going to create dozens / hundreds of views in the upcoming months/years, and as they can be dependent on each other, what is the best practice to use since schemachange uses a simple alphabetical order for exectuions. 

e.g.: if I  have the following views: 
R__1___view1.sql
R__2___view2.sql
R__3___view3.sql

this is all good until I need to make a change e.g.: to "view1" which makes it dependent on "view3" or a new view, e.g.: "view4"
At this point, to keep this runnable, we'd need to renumber all scripts as follows: 

R__1___view3.sql
R__2___view4.sql
R__3___view1.sql
R__4___view2.sql

Which can become a problem when talking dozens or hundreds of views. 

I was thinking about using a naming convention like the following: 
R__5100_view1
R__5200_view2
R__5300_view3

This way I can simply add "view4" in one of the gaps and rename "view3" to use another slot of the gaps. 
R__5040_view3
R__5050_view4
R__5100_view1
R__5200_view2

What is the best way of handling this? 
I was thinking about a file to track dependencies, essentially putting all the scripts in the order of desired execution. 
a bit of extra overhead but would still be a lot simpler then messing with the numbering / renaming exercise. 

Best,
Gergely

**Comment by carter293** [comment]
I would also appreciate support for defining dependency.

**Comment by sfc-gh-tmathew** [comment]
@carter293 @GergelyGabor 

Thanks for the question! Here's how you can manage dependencies with schemachange:

## Organizing Your Scripts

The key is to organize your migration scripts into separate folders based on their dependencies:

1. **Independent objects** - Put these in one folder (e.g., `migrations/01_base/`)
2. **Dependent objects** - Organize these in separate folders following the order of dependency (e.g., `migrations/02_dependent/`, `migrations/03_more_dependent/`)

Use a naming convention that works for your team and clearly indicates the execution order.

## Executing in Order

Run schemachange multiple times in your workflow, passing the `-f` or `--root-folder` argument for each step:

schemachange deploy -f migrations/01_base/
schemachange deploy -f migrations/02_dependent/
schemachange deploy -f migrations/03_more_dependent/

This ensures scripts are executed in the correct dependency order.

PS: This pattern is not useful when you have a monorepo with different cadence of releases. You will need to separate your repositories into separate repos so that cross project dependencies and coordinated and not just dependenent on tools.

## Future Considerations

Unlike dbt, schemachange doesn't currently have DAG-like dependency management built-in. We're aware this would be a valuable feature and will evaluate the effort required to add something like this in a future release. For now, the folder-based approach above should give you good control over execution order.

Hope this helps!
